<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.280">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Malte D Luecken">
<meta name="author" content="Maren Buttner">
<meta name="author" content="Kridsadakorn Chaichoompu">
<meta name="author" content="Anna Danese">
<meta name="author" content="Marta Interlandi">
<meta name="author" content="Michaela F Mueller">
<meta name="author" content="Daniel C Strobl">
<meta name="author" content="Luke Zappia">
<meta name="author" content="Martin Dugas">
<meta name="author" content="Maria Colome-Tatche">
<meta name="author" content="Fabian J Theis">
<meta name="dcterms.date" content="2020-05-23">

<title>quarto-blog - Benchmarking atlas-level data integration in single-cell genomics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">quarto-blog</span>
    </a>
  </div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../posts/index.html">
 <span class="menu-text">Posts</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../projects/index.html">
 <span class="menu-text">Projects</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../talks/index.html">
 <span class="menu-text">Talks</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../publications/index.html">
 <span class="menu-text">Publications</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/"><i class="bi bi-github" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com"><i class="bi bi-twitter" role="img">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
              <div id="quarto-search" class="" title="Search"></div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Benchmarking atlas-level data integration in single-cell genomics</h1>
  <div class="quarto-categories">
    <div class="quarto-category">single-cell</div>
    <div class="quarto-category">rna-seq</div>
    <div class="quarto-category">integration</div>
    <div class="quarto-category">benchmarking</div>
  </div>
  </div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Authors</div>
    <div class="quarto-title-meta-contents">
             <p>Malte D Luecken </p>
             <p>Maren Buttner </p>
             <p>Kridsadakorn Chaichoompu </p>
             <p>Anna Danese </p>
             <p>Marta Interlandi </p>
             <p>Michaela F Mueller </p>
             <p>Daniel C Strobl </p>
             <p><strong>Luke Zappia</strong> </p>
             <p>Martin Dugas </p>
             <p>Maria Colome-Tatche </p>
             <p>Fabian J Theis </p>
          </div>
  </div>
  
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">May 23, 2020</p>
    </div>
  </div>
  
  
    <div>
    <div class="quarto-title-meta-heading">Links</div>
    <div class="quarto-title-meta-contents">
              <a type="button" class="btn btn-outline-primary" href="https://www.biorxiv.org/content/10.1101/2020.05.22.111161v1.full.pdf">PDF</a>
                    <a type="button" class="btn btn-outline-primary" href="https://github.com/theislab/scib">Package</a>
                    <a type="button" class="btn btn-outline-primary" href="https://github.com/theislab/scib-reproducibility">Code</a>
                    <a type="button" class="btn btn-outline-primary" href="https://doi.org/10.1101/2020.05.22.111161">DOI</a>
          </div>
  </div>
  
    <div>
    <div class="quarto-title-meta-heading">Citation stats</div>
    <div class="quarto-title-meta-contents">
      <div class="scite-badge" data-doi="10.1101/2020.05.22.111161" data-layout="vertical" data-show-zero="true" data-small="false" data-show-labels="true" data-tally-show="true">
      </div>
    </div>
  </div>
  <script async="" type="application/javascript" src="https://cdn.scite.ai/badge/scite-badge-latest.min.js">
  </script>
  
</div>


<div>
  <div class="abstract">
    <div class="abstract-title">Abstract</div>
    <p>Cell atlases often include samples that span locations, labs, and conditions, leading to complex, nested batch effects in data. Thus, joint analysis of atlas datasets requires reliable data integration. Choosing a data integration method is a challenge due to the difficulty of defining integration success. Here, we benchmark 38 method and preprocessing combinations on 77 batches of gene expression, chromatin accessibility, and simulation data from 23 publications, altogether representing &gt;1.2 million cells distributed in nine atlas-level integration tasks. Our integration tasks span several common sources of variation such as individuals, species, and experimental labs. We evaluate methods according to scalability, usability, and their ability to remove batch effects while retaining biological variation. Using 14 evaluation metrics, we find that highly variable gene selection improves the performance of data integration methods, whereas scaling pushes methods to prioritize batch removal over conservation of biological variation. Overall, BBKNN, Scanorama, and scVI perform well, particularly on complex integration tasks; Seurat v3 performs well on simpler tasks with distinct biological signals; and methods that prioritize batch removal perform best for ATAC-seq data integration. Our freely available reproducible python module can be used to identify optimal data integration methods for new data, benchmark new methods, and improve method development.</p>
  </div>
</div>

</header>




<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents"><h2 class="anchored quarto-appendix-heading">Citation</h2><div><div class="quarto-appendix-secondary-label">BibTeX citation:</div><pre class="sourceCode code-with-copy quarto-appendix-bibtex"><code class="sourceCode bibtex">@misc{dluecken2020,
  author = {Malte D Luecken and Maren Buttner and Kridsadakorn
    Chaichoompu and Anna Danese and Marta Interlandi and Michaela F
    Mueller and Daniel C Strobl and **Luke Zappia** and Martin Dugas and
    Maria Colome-Tatche and Fabian J Theis},
  title = {Benchmarking Atlas-Level Data Integration in Single-Cell
    Genomics},
  pages = {2020.05.22.111161},
  date = {2020-05-23},
  doi = {10.1101/2020.05.22.111161},
  langid = {en},
  abstract = {Cell atlases often include samples that span locations,
    labs, and conditions, leading to complex, nested batch effects in
    data. Thus, joint analysis of atlas datasets requires reliable data
    integration. Choosing a data integration method is a challenge due
    to the difficulty of defining integration success. Here, we
    benchmark 38 method and preprocessing combinations on 77 batches of
    gene expression, chromatin accessibility, and simulation data from
    23 publications, altogether representing \textgreater1.2 million
    cells distributed in nine atlas-level integration tasks. Our
    integration tasks span several common sources of variation such as
    individuals, species, and experimental labs. We evaluate methods
    according to scalability, usability, and their ability to remove
    batch effects while retaining biological variation. Using 14
    evaluation metrics, we find that highly variable gene selection
    improves the performance of data integration methods, whereas
    scaling pushes methods to prioritize batch removal over conservation
    of biological variation. Overall, BBKNN, Scanorama, and scVI perform
    well, particularly on complex integration tasks; Seurat v3 performs
    well on simpler tasks with distinct biological signals; and methods
    that prioritize batch removal perform best for ATAC-seq data
    integration. Our freely available reproducible python module can be
    used to identify optimal data integration methods for new data,
    benchmark new methods, and improve method development.}
}
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre><div class="quarto-appendix-secondary-label">For attribution, please cite this work as:</div><div id="ref-dluecken2020" class="csl-entry quarto-appendix-citeas" role="doc-biblioentry">
Malte D Luecken, Maren Buttner, Kridsadakorn Chaichoompu, Anna Danese,
Marta Interlandi, Michaela F Mueller, Daniel C Strobl, et al. 2020.
<span>“Benchmarking Atlas-Level Data Integration in Single-Cell
Genomics.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2020.05.22.111161">https://doi.org/10.1101/2020.05.22.111161</a>.
</div></div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>