---
title: Bioconductor 3.3 packages
author: Luke Zappia
date: '2016-05-03'
slug: bioconductor-3-3-packages
categories:
  - migrated
tags:
  - R
  - Bioconductor
---

Bioconductor 3.3 has just been released. You can find the complete list of new
packages (and changes to existing packages) 
[here](https://bioconductor.org/news/bioc_3_3_release/) but here are a few I
thought might be interesting based on the description. I might have more to say
once I've had time to try a few out.

* **debrowser** – Interactive plots and tables for differential expression
* **DEFormats** – convert between differential expression formats
* **EBSEA** – exon based differential expression
* **EmpiricalBrownsMethod** – combining dependent p-values
* **Linnorm** – normalisation for parametric tests, simulation of RNA-seq data
* **multiClust** – feature selection and clustering analysis for transcriptomic
  data
* **RGraph2js** – interactive network visualisations with D3
* **tximport** – import and summarise transcript-level estimates

## Single-cell

These packages are specific to single-cell RNA-seq analysis. A couple of them I
am already familiar with, particularly **scater**.

* **cellity** - identifying low-quality cells
* **celTree** - model the relationship between individual cells over time or
  space.
* **scater** - tools for analysis of single-cell RNA-seq data (particularly QC)
* **scde** - single-cell differential expression
* **scran** - normalisation, cell-cycle assignment, gene detection
