---
title: Benchmarking atlas-level data integration in single-cell genomics
date: 2020-05-23T16:02:00
author:
 - Malte D Luecken
 - Maren Buttner
 - Kridsadakorn Chaichoompu
 - Anna Danese
 - Marta Interlandi
 - Michaela F Mueller
 - Daniel C Strobl
 - "**Luke Zappia**"
 - Martin Dugas
 - Maria Colome-Tatche
 - Fabian J Theis
abstract: |
    Cell atlases often include samples that span locations, labs, and conditions, leading to complex, nested batch effects in data.
    Thus, joint analysis of atlas datasets requires reliable data integration.
    Choosing a data integration method is a challenge due to the difficulty of defining integration success.
    Here, we benchmark 38 method and preprocessing combinations on 77 batches of gene expression, chromatin accessibility, and simulation data from 23 publications, altogether representing >1.2 million cells distributed in nine atlas-level integration tasks.
    Our integration tasks span several common sources of variation such as individuals, species, and experimental labs.
    We evaluate methods according to scalability, usability, and their ability to remove batch effects while retaining biological variation.
    Using 14 evaluation metrics, we find that highly variable gene selection improves the performance of data integration methods, whereas scaling pushes methods to prioritize batch removal over conservation of biological variation.
    Overall, BBKNN, Scanorama, and scVI perform well, particularly on complex integration tasks; Seurat v3 performs well on simpler tasks with distinct biological signals; and methods that prioritize batch removal perform best for ATAC-seq data integration.
    Our freely available reproducible python module can be used to identify optimal data integration methods for new data, benchmark new methods, and improve method development.
categories: ["single-cell", "rna-seq", "integration", "benchmarking"]
links:
  doi: "10.1101/2020.05.22.111161"
  pdf: "https://www.biorxiv.org/content/10.1101/2020.05.22.111161v1.full.pdf"
  package: "https://github.com/theislab/scib"
  code: "https://github.com/theislab/scib-reproducibility"
citation:
  type: article
  doi: 10.1101/2020.05.22.111161
  container-title: bioRxiv
  page: 2020.05.22.111161
---
