---
title: "Exploring the single-cell RNA-seq analysis landscape with the scRNA-tools database"
date: 2018-06-01
publishDate: 2020-03-14T13:18:37.645705Z
authors: ["**Luke Zappia**", "Belinda Phipson", "Alicia Oshlack"]
publication_types: ["2"]
abstract: "As single-cell RNA-sequencing (scRNA-seq) datasets have become more widespread the number of tools designed to analyse these data has dramatically increased. Navigating the vast sea of tools now available is becoming increasingly challenging for researchers. In order to better facilitate selection of appropriate analysis tools we have created the scRNA-tools database (www.scRNA-tools.org) to catalogue and curate analysis tools as they become available. Our database collects a range of information on each scRNA-seq analysis tool and categorises them according to the analysis tasks they perform. Exploration of this database gives insights into the areas of rapid development of analysis methods for scRNA-seq data. We see that many tools perform tasks specific to scRNA-seq analysis, particularly clustering and ordering of cells. We also find that the scRNA-seq community embraces an open-source and open-science approach, with most tools available under open-source licenses and preprints being extensively used as a means to describe methods. The scRNA-tools database provides a valuable resource for researchers embarking on scRNA-seq analysis and records the growth of the field over time."
featured: false
publication: "*PLoS computational biology*"
tags: ["single-cell", "rna-seq", "database"]
url_pdf: "https://journals.plos.org/ploscompbiol/article/file?id=10.1371/journal.pcbi.1006245&type=printable"
url_code: "https://github.com/Oshlack/scRNA-tools"
url_project: "https://www.scrna-tools.org/"
url_source: "https://github.com/Oshlack/scRNAtools-paper"
url_preprint: "https://doi.org/10.1101/206573"
doi: "10.1371/journal.pcbi.1006245"
---
